lib/pure/times.nim

```diff
@@ -1396,228 +1488,725 @@ proc `*=`*[T: TimesMutableTypes, U](a: var T, b: U) =
+    h, hh, H, HH
+    m, mm, M, MM, MMM, MMMM
+    s, ss
+    fff, ffffff, fffffffff
+    t, tt
+    y, yy, yyy, yyyy, yyyyy
+    YYYY
+    uuuu
+    UUUU
+    z, zz, zzz, zzzz
+    g
+
+    # This is a special value used to mark literal layout values.
+    # See the doc comment for ``TimeLayout.patterns``.
+    Lit
+
+  TimeLayout* = object ## Represents a format for parsing and printing
+                       ## time types.
+    patterns: seq[byte] ## \
+      ## Contains the patterns encoded as bytes.
+      ## Literal values are encoded in a special way.
+      ## They start with ``Lit.byte``, then the length of the literal, then the
+      ## raw char values of the literal. For example, the literal `foo` would
+      ## be encoded as ``@[Lit.byte, 3.byte, 'f'.byte, 'o'.byte, 'o'.byte]``.
+    layout: string
+
+const PatternSeperators = { ' ', '-', '/', ':', '(', ')', '[', ']', ',' }
+
+proc `$`*(f: TimeLayout): string =
+  ## Returns the layout that was used to construct ``f``.
   runnableExamples:
-    let dt = initDateTime(01, mJan, 2000, 12, 00, 00, 01, utc())
-    doAssert format(dt, "yyyy-MM-dd'T'HH:mm:ss'.'fffffffffzzz") == "2000-01-01T12:00:00.000000001+00:00"
+    let f = initTimeLayout("yyyy-MM-dd")
+    doAssert $f == "yyyy-MM-dd"
+  f.layout
 
-  result = ""
+proc raiseParseException(f: TimeLayout, input: string, msg: string) =
+  raise newException(ValueError,
+                     "Failed to parse date '" & input & "' with " &
+                     "format '" & $f & "'. " & msg)
+
+iterator tokens(f: string): tuple[kind: TokenKind, token: string] =
+  var formattedToken: string
   var i = 0
-  var currentF = ""
-  while i < f.len:
-    case f[i]
-    of ' ', '-', '/', ':', '\'', '(', ')', '[', ']', ',':
-      formatToken(dt, currentF, result)
+  var currToken = ""
 
-      currentF = ""
+  template yieldcurrToken() =
+    if currToken.len != 0:
+      yield (tkPattern, currToken)
+      currToken = ""
 
-      if f[i] == '\'':
+  while i < f.len:
+    case f[i]
+    of '\'':
+      yieldcurrToken()
+      if f[i.succ] == '\'':
+        yield (tkLiteral, "'")
+        i.inc 2
+      else:
+        var token = ""
         inc(i) # Skip '
-        while i < f.len-1 and f[i] != '\'':
-          result.add(f[i])
-          inc(i)
-      else: result.add(f[i])
-
+        while f[i] != '\'' and i < f.high:
+          token.add f[i]
+          i.inc
+        i.inc
+        yield (tkLiteral, token)
+    of PatternSeperators:
+        yieldcurrToken()
+        yield (tkLiteral, $f[i])
+        i.inc
     else:
       # Check if the letter being added matches previous accumulated buffer.
-      if currentF.len == 0 or currentF[high(currentF)] == f[i]:
-        currentF.add(f[i])
+      if currToken.len == 0 or currToken[0] == f[i]:
+        currToken.add(f[i])
+        i.inc
+      else:
+        yield (tkPattern, currToken)
+        currToken = $f[i]
+        i.inc
+
+  yieldcurrToken()
+
+proc stringToPattern(str: string): Pattern =
```

Comment:
Couldn't parseEnum be used for this?

Reply:
`parseEnum` is completely case insensitive for some reason